hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/EncryptedFileSystem.java

```diff
@@ -145,7 +145,7 @@ public boolean rename(Path src, Path dst) throws IOException {
 
         if (renamed) {
             tryRemoveKey(src);
-        } else {
+        } else if (!fs.exists(dst)) {
```

Comment:
I think we need to do this `fs.exists(dst)` check at the top of the method. At this point `keyStore.put(...)` has already been called and would have overwritten the key material.

Reply:
I can add the check at the top! That said, we'd also want it here unfortunately. As I call out in the PR unfortunately, there are two edge cases for `rename` - one where the file already exists in the destination, and another where rename does properly create the file in the new location, but it still returns false.

Reply:
Ok so the check at the top deal with the first edge case (file already exists at the destination and we exit immediately).

How can the second edge case happen? Like `fs.rename(src, dst)` actually succeeds but then return `false` even though it was successful?

Reply:
I guess the other thing is that if a file already exists at dst I feel like we should be relying on the `fs.rename` behavior? Which either should fail because a file already exists, at which point I'd expect `tryRemoveKey(dst)` to be executed, or it actually overwrites the dst file in which case rename success and we run `tryRemoveKey(src)` which seems right?

Reply:
After collecting some more data, I've confirmed that hadoop-azure can run `rename`, return false indicating that it failed, but it leaves around cruft: in addition to it properly leaving the source file around, the destination file can exist too. And that destination file might only be partially written to, or it could even have the same file size as the source. I verified that when we've seen this behavior, the destination file didn't exist before calling `rename` but as mentioned it existed after calling it, even if it fails.

Going to solve this in client code instead for simplicity.